File: contest/ABC324/C.jl

# S の方が長さが 1 だけ長いとき、　一回スキップありで diff を調べる
function skip_diff(S, T)
    S_idx = 1
    T_idx = 1
    use_skip = false
    N_T = length(T)
    while true
        if S[S_idx] != T[T_idx]
            if use_skip
                return false
            else
                S_idx += 1
                use_skip = true
            end
        else
            if T_idx == N_T 
                return true
            end
            S_idx += 1
            T_idx += 1
        end
    end
    return false
end


function ndiff(S, T)
    n_diff = 0
    for i in eachindex(S)
        if S[i] != T[i]
            n_diff += 1
        end
    end
    return n_diff
end




function main()
    N, T = split(readline())
    N = parse(Int, N)

    result = Int[]

    for i in 1:N

        S = readline()


        if abs(length(S) - length(T)) > 1
            continue
        end

        # 同一
        if S == T
            push!(result, i)
            continue
        end

        # Sに一文字追加で同じになるか判定
        if length(T) - length(S) == 1
            if skip_diff(T, S)
                push!(result, i)
                continue
            end
        end


        # Sを一文字削除して同じになるか判定。これはさっきのやつを反転させればいい
        if length(S) - length(T) == 1
            if skip_diff(S, T)
                push!(result, i)
                continue
            end
        end

        # 一文字置換
        if length(S) == length(T)
            if ndiff(S, T) == 1
                push!(result, i)
            end
            continue
        end
    end


    K = length(result)

    println(K)

    S = ""

    for i in 1:K
        S *= repr(result[i]) * " "
    end

    println(S)

end


main()

File: contest/ABC324/C.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

bool skip_diff(string s, string t){
    int S_idx = 0;
    int T_idx = 0;
    bool use_skip  = false;
    int N_T = t.size();
    while (true) {
        if (s.at(S_idx) != t.at(T_idx)) {
            if (use_skip) {
                return false;
            }
            use_skip = true;
            S_idx++;
        }
        else {
            if (T_idx == N_T - 1) {
                return true;
            }
            S_idx++;
            T_idx++;
        }
    }
}


int ndiff(string s, string t){
    int n_diff = 0;
    for (int i = 0; i < s.size(); i++)
    {
        if (s.at(i) != t.at(i)) {
            n_diff++;
        }
    }
    return n_diff;
}


int main(){
    std::cin.tie(0)->sync_with_stdio(0);
    int N;
    string T;
    cin >> N >> T;

    int K = 0;
    vector<int> ans;

    for (int i = 0; i < N; i++)
    {
        string S;
        cin >> S;
        if (S.size() == T.size()) {
            if (ndiff(S, T) <= 1) {
                K++;
                ans.push_back(i);
            }
        }
        else if (S.size() == T.size() + 1) {
            if (skip_diff(S, T)) {
                K++;
                ans.push_back(i);
            }
        }
        else if (S.size() == T.size() - 1) {
            if (skip_diff(T, S)) {
                K++;
                ans.push_back(i);
            }
        }
    }

    cout << K << endl;
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans.at(i) + 1 << " ";
    }

    std::cout << std::endl;

    return 0;
}
